Docs for `JupyterAnalysis` and `Action` schemas (#77)
FAIRmat-NFDI/nomad-analysis: docs/how_to/run_nomad_actions_from_eln.md

# Run NOMAD Actions from the ELN

[NOMAD Actions][ext_href_1] allow to define executable workflows that can run on
specialized workers. This guide explains how to use the `Action` schema from
`nomad_analysis.actions.schema` to trigger, monitor, and stop NOMAD Actions

> directly from the ELN interface.

Reviewer: Wow, didn't know that triggering also goes in this direction. Awesome!

If you would like to set up customizable analysis workflows that run in Jupyter
notebooks instead, check out the [Analysis with Jupyter Notebooks][int_href_1]

> guide.

Reviewer: This is a prerequisite, no?

Author: No, Actions can be set and run completely independently of JupyterAnalysis

!!! tip "When to use Actions over Jupyter Notebooks?"
    - Running standarized analysis workflows that require fixed user input and generates a fixed output, without any custom code
    - Offloading heavy computations to dedicated workers

>     - Integrate with external services and APIs as part of your analysis

Reviewer: Would move this into your intro paragraph above.

## Overview

The `Action` base schema from `nomad_analysis.actions.schema` module is a
reusable schema for integrating Actions with NOMAD's ELN interface.
It provides a boilerplate NOMAD section equipped with buttons to:

- **Start Action**: Trigger a workflow from the ELN
- **Get Status**: Check the current status of a running action
- **Stop Action**: Cancel a running action

> The schema is designed to be extended for different types of analysis workflows. By implementing the `start_action` method, you can define custom logic to prepare inputs and trigger any registered NOMAD Action.

Reviewer: > you can [...] trigger any registered NOMAD Action.

What does that mean? That you can control when Actions are deployed automatically, e.g. during normalization? If so, how are events denoted?

Author: Yeah, from these ELN schemas, one can trigger Actions. However, that's not the only way to trigger them. In the new GUI, they can be triggered from forms that appear as Dialog boxes.

!!! hint
    Read more about registration of Actions in the

>     [Action Entry Point][ext_href_2] guide.

Reviewer: Would merge with the previous paragraph.

## Create a custom Action Schema

Extend the `Action` class and implement the `start_action` method:

```python
from nomad.actions import manager
from nomad.metainfo import Quantity, Section
from nomad_analysis.actions.schema import Action

# import your custom action input schema if needed
from .models import MyAnalysisActionInput

class MyAnalysisAction(Action):
    """Custom action for running analysis workflows."""

    # Add custom quantities for your action inputs
    input_parameter = Quantity(
        type=str,
        description='Input parameter for the analysis.',

>         a_eln={'component': 'StringEditQuantity'},

Reviewer: Properties should always be editable to interact with an analysis, is that right? At least the results should be editable? Again, very technical question, but I keep seeing this pop up in your schema.

Author: Yeah, string edit quantity makes them editable by default.

    )

    def start_action(self, archive, logger) -> str:
        """
        Triggers the analysis action.

        Returns:
            str: The instance ID of the triggered action.
        """
        # Prepare input for the action
        action_input = MyActionInput(
            user_id=archive.metadata.authors[0].user_id,
            upload_id=archive.metadata.upload_id,
            parameter=self.input_parameter,

>         )

Reviewer: Does this mean that the action triggers once all these fields have been filled?

Author: Yes, the input to the actions is based on some BaseModel, which needs to be validated before the action is started

        # Start the action using the NOMAD action manager
        instance_id = manager.start_action(

>             action_id='nomad_example.actions.myaction:my_action',

Reviewer: This ID is tracked in the logs?

Author: Usually, this id is hard coded into a given ELN schema, that is meant to trigger one particular action. We are not logging this. In the entry, we are preserving the action **instance** ID, which is unique for every run

            data=action_input,
        )

        return instance_id
```

## Use in an ELN Entry

The custom action section can be added as a sub-section to an ELN schema:

```python
from nomad.datamodel.data import EntryData
from nomad.metainfo import SubSection

class MyAnalysisELN(EntryData):
    """ELN entry with action trigger capability."""

    action = SubSection(section_def=MyAnalysisAction)
```

Or it can be defined as a standalone ELN entry:

```python
from nomad.datamodel.data import EntryData

class MyAnalysisELN(MyAnalysisAction, EntryData):
    """ELN entry for my specific analysis workflow."""

> ```

Reviewer: So what? Sections are naturally recursive in NOMAD... If you want to point out that it does not have to be the top section, sure, but a single sentence suffices.

Author: Made it more clear

## Troubleshooting

If something isn't working as expected, there's a good chance you will find

> useful information in the entry processing logs. You can access these logs from the ELN interface. Some common issues and tips:

Reviewer: No verb abbreviation in formal texts:

```suggestion
If something does not work as expected, check the entry processing logs first.
These logs are accessible from the ELN interface <explanation on how to get there>.
Some common issues and tips:
```

> - **Action Won't Start**: Check if an action is already running. Only one action can run at a time per entry. Either stop the running action or wait for it to complete before starting a new one.

Reviewer: ```suggestion
- **Action Won't Start**: Check whether an action is already running. Only one action can run at a time per entry. Either stop the running action or wait for it to complete before starting a new one.
```

> - **Status Not Updating**: The status doesn't auto-refresh. You need to click "Get Action Status" to see the latest status. If it still doesn't update, check the logs for errors.

Reviewer: ```suggestion
- **Status Not Updating**: The status does not auto-refresh. Try clicking "Get Action Status" to manually retrieve the latest status. If the status does not update, check the logs for errors.
```
